File: ruby/scrape.rb
# Fetch a page through the ScrapingBee web scraping API.
# Auth is the Authorization: Bearer header. The api_key query parameter is deprecated.

require 'net/http'
require 'uri'

API_URL = 'https://app.scrapingbee.com/api/v1'.freeze
API_KEY = ENV.fetch('SCRAPINGBEE_API_KEY', 'YOUR-API-KEY')

def scrape(target, params = {})
  uri = URI(API_URL)
  uri.query = URI.encode_www_form({ url: target }.merge(params))

  request = Net::HTTP::Get.new(uri)
  request['Authorization'] = "Bearer #{API_KEY}"

  Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
    http.request(request)
  end
end

# render_js=false keeps this call at 1 credit instead of the 5-credit default.
response = scrape('https://news.ycombinator.com', render_js: 'false')

puts "status: #{response.code}"
puts "credits: #{response['Spb-cost']}"
puts "request id: #{response['Spb-request-id']}"
puts response.body[0, 500]

File: go/scrape.go
// Fetch a page through the ScrapingBee web scraping API.
// Auth is the Authorization: Bearer header. The api_key query parameter is deprecated.
package main

import (
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
)

const apiURL = "https://app.scrapingbee.com/api/v1"

func scrape(target string, params map[string]string) (*http.Response, error) {
	apiKey := os.Getenv("SCRAPINGBEE_API_KEY")
	if apiKey == "" {
		apiKey = "YOUR-API-KEY"
	}

	query := url.Values{}
	query.Set("url", target)
	for key, value := range params {
		query.Set(key, value)
	}

	req, err := http.NewRequest("GET", apiURL+"?"+query.Encode(), nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)

	return (&http.Client{}).Do(req)
}

func main() {
	// render_js=false keeps this call at 1 credit instead of the 5-credit default.
	resp, err := scrape("https://news.ycombinator.com", map[string]string{"render_js": "false"})
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("status:", resp.Status)
	fmt.Println("credits:", resp.Header.Get("Spb-cost"))
	fmt.Println("request id:", resp.Header.Get("Spb-request-id"))
	fmt.Println(string(body[:min(500, len(body))]))
}
